### PR TITLE
CDAP-6589 Support Impersonation for logs

### PR DESCRIPTION
**Pull request by rohitsinha54**
Issue: https://issues.cask.co/browse/CDAP-6589
Build: http://builds.cask.co/browse/CDAP-DUT4476-1
- Supports impersonation for logs

Note: 
- This needs to be merged against develop. Opening against another feature branch for review. 
- This needs to be tested on a cluster.

### CODE REVIEW

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/write/LogCleanup.java [hunk 1]
```diff
@@ -57,28 +64,45 @@ public void run() {
     try {
       long tillTime = System.currentTimeMillis() - retentionDurationMs;
       final SetMultimap<String, Location> parentDirs = HashMultimap.create();
+      final Map<String, NamespaceId> namespacedLogBaseDirMap = new HashMap<>();
       fileMetaDataManager.cleanMetaData(tillTime,
                                         new FileMetaDataManager.DeleteCallback() {
                                           @Override
-                                          public void handle(Location location, String namespacedLogBaseDir) {
```
Comment (anwar6953) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/write/LogCleanup.java:63]: Where is this Location object created? We only need to wrap the creation of the Location within to `doAs`. The operations on that Location don't have to be in the `doAs` block.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/write/LogCleanup.java [hunk 2]
```diff
@@ -57,28 +64,45 @@ public void run() {
...
       // Delete any empty parent dirs
-      for (String namespacedLogBaseDir : parentDirs.keySet()) {
-        Set<Location> locations = parentDirs.get(namespacedLogBaseDir);
-        for (Location location : locations) {
-          deleteEmptyDir(namespacedLogBaseDir, location);
-        }
+      for (final String namespacedLogBaseDir : parentDirs.keySet()) {
+        // this ensures that we only do doAs which will make an RPC call only once for a namespace
+        NamespaceId namespaceId = namespacedLogBaseDirMap.get(namespacedLogBaseDir);
+        impersonator.doAs(namespaceId, new Callable<Void>() {
```
Comment (anwar6953) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/write/LogCleanup.java:96]: we don't need to wrap these operations within a `doAs`. Just the `Location` object has to be created within a `doAs`.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/file/FileLogAppender.java [hunk 3]
```diff
@@ -82,13 +84,14 @@ public FileLogAppender(CConfiguration cConfig,
                          DatasetFramework dsFramework,
                          TransactionExecutorFactory txExecutorFactory,
                          NamespacedLocationFactory namespacedLocationFactory,
-                         RootLocationFactory rootLocationFactory) {
+                         RootLocationFactory rootLocationFactory, Impersonator impersonator) {
```
Comment (anwar6953) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/file/FileLogAppender.java:87]: My understanding is that FileLogAppender is only used in Standalone (not in distributed).
Impersonation is not supported in Standalone.

**Comment by anwar6953** [comment]
Change also needs to happen in the log handlers right?

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/write/LogCleanup.java (hunk 1 above)
Comment (rohitsinha54) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/write/LogCleanup.java:63]: @anwar6953 My understanding from what was conveyed earlier was that the operation such as reading/writing/deleting/exists etc needs to be wrapped. Your comment says something else now. Can you confirm what exactly needs to be done ? It will be better if you add the requirement in the jira. Also, is this change required due to this issue: https://issues.cask.co/browse/CDAP-6606

**Comment by rohitsinha54** [comment]
Ran cdap-watchdog module tests locally.

**Comment by rohitsinha54** [comment]
@anwar6953 : Why do we need in handler. The handler calls end up the the FileReader itself.

**Comment by rohitsinha54** [comment]
Build: http://builds.cask.co/browse/CDAP-DUT4476-3
ITN: http://builds.cask.co/browse/CDAP-ITM-24

**Comment by anwar6953** [comment]
Right - we won't need the impersonation in the handler, if its being done in the underlying FileReader class.

**Comment by rohitsinha54** [comment]
closing this. opened against develop https://github.com/caskdata/cdap/pull/6164
